src/zps.c

```diff
@@ -15,439 +15,797 @@
-static char *strPath;                    /* String part of a path in '/proc' */
-static char *statContent;                /* Text content of the process's stat file */
-static char *cmdContent;                 /* Text content of the process's command file */
-static char *procEntryColor;             /* Color code of process entry to print */
-typedef struct {                         /* Struct for storing process stats */
-    unsigned int pid;
-    unsigned int ppid;
-    char name[BLOCK_SIZE/64];
-    char state[BLOCK_SIZE/64];
-    char cmd[BLOCK_SIZE];
-    bool defunct;
-} ProcStats;
-static ProcStats
-    defunctProcs[BLOCK_SIZE/4]; /* Array of defunct process's stats */
-static va_list vargs;           /* List of information about variable arguments */
-static regex_t regex;           /* Regex struct */
-static regmatch_t               /* Regex match struct that contains start and end offsets */
-    regMatch[REG_MAX_MATCH];
-static struct option opts[] = { /* Long options for command line arguments  */
-    {"version", no_argument,
-        NULL, 'v'},
-    {"help", no_argument,
-        NULL, 'h'},
-    {"reap", no_argument,
-        NULL, 'r'},
-    {"lreap", no_argument,
-        NULL, 'x'},
-    {"list", no_argument,
-        NULL, 'l'},
-    {"prompt", no_argument,
-        NULL, 'p'},
-    {"silent", no_argument,
-        NULL, 's'},
-    {"fd", required_argument,
-        NULL, 'f'},
+#include <unistd.h>
+
+#include "zps.h"
+
+static const char *const abbrevs[NSIG] = {
+    [SIGHUP] = "HUP",       [SIGINT] = "INT",     [SIGQUIT] = "QUIT",
+    [SIGILL] = "ILL",       [SIGTRAP] = "TRAP",   [SIGABRT] = "ABRT",
+    [SIGFPE] = "FPE",       [SIGKILL] = "KILL",   [SIGBUS] = "BUS",
+    [SIGSYS] = "SYS",       [SIGSEGV] = "SEGV",   [SIGPIPE] = "PIPE",
+    [SIGALRM] = "ALRM",     [SIGTERM] = "TERM",   [SIGURG] = "URG",
+    [SIGSTOP] = "STOP",     [SIGTSTP] = "TSTP",   [SIGCONT] = "CONT",
+    [SIGCHLD] = "CHLD",     [SIGTTIN] = "TTIN",   [SIGTTOU] = "TTOU",
+    [SIGPOLL] = "POLL",     [SIGXCPU] = "XCPU",   [SIGXFSZ] = "XFSZ",
+    [SIGVTALRM] = "VTALRM", [SIGPROF] = "PROF",   [SIGUSR1] = "USR1",
+    [SIGUSR2] = "USR2",     [SIGWINCH] = "WINCH",
 };
 
 /*!
- * Write colored and formatted data to stream. (stderr)
+ * Helper to get the string abbreviation of signal constants
+ *
+ * @param[in] sig Signal number to get the string representation of
+ *
+ * @return String representing the signal constant (abbreviated),
+ *         or NULL if no corresponding signal string was found
+ */
+static const char *sig_abbrev(int sig)
+{
+    if (sig < 0 || !((size_t)sig < sizeof(abbrevs))) {
+        return NULL;
+    }
+    return abbrevs[sig];
+}
+
+/*!
+ * Attempts to find the corresponding signal number to the
+ * given signal constant's name
+ *
+ * @param[in] sig_str Signal name to convert
+ *
+ * @return -1 on error, the corresponding signal number otherwise
+ */
+static int sig_str_to_num(const char *sig_str)
+{
+    assert(sig_str);
```

Comment:
I've seen these `assert`s throughout the code I'm assuming these are for diagnostics, right?

Reply:
Yes, these runtime assertions are simply there for testing and debugging purposes, checking the preconditions of the functions which should hold at all times – for example, we do not intend to pass `NULL` pointers. In the releases, we can disable those runtime assertions at compile time by defining `NDEBUG`, as was added to the CMake configuration and the Makefile. The testing container should still have the assertions enabled, though.

Reply:
Got it! Do you think we should disable them for the container as well?

Reply:
We probably should do so in `docker/Dockerfile`, but not in `test.sh`. I will add the flag to the Dockerfile for this PR.